### DISCUSSION TITLE
Debugging lingering batches

### DISCUSSION THREAD

**Discussion by dhughes** [sidekiq/sidekiq#5848] [Q&A]
I see the documentation for batches says this:

> Pending batches will expire in Redis after 30 days. Callbacks won't trigger and you will have to deal with performing any cleanup work manually.

I have a lot of batches that are in a weird state. They go back as far as 4 weeks ago. Here's a screenshot of a few of these:

![Screenshot 2023-03-24 at 7 17 54 AM](https://user-images.githubusercontent.com/573772/227507620-723bac10-af79-48b4-ad9e-31389227fc35.png)

I'm having a hard time figuring out what's going on via this UI. For example, the first batch in the screenshot shows 1 job count, 1 pending, and 1 failed. Is that trying to tell me that the failed job is enqueued to be retried? If so... what happened to the job? I only retry 8 times, so if that job retried until death, it would have only taken about 1.5 hours. So, it seems weird that this batch is still here. 

The other examples also seem weird. The one with 0 jobs (which I know is undefined) must presumably be a case where I've created a batch, but an error was raised before I could add any jobs to it. I'll have to see if I can figure that one out. And the one that's all green with 0 pending and 0 errors. Any idea why that might happen?

Also, sometimes I see the following appearing in Bugsnag. None of this is my code and I haven't been able to figure out what's up. Any chance it's related?

```
TypeError
Sidekiq::Batch::Callback@default
no implicit conversion of nil into String
Mar 18th 2023, 15:40:06 EDT

STACKTRACE

TypeError no implicit conversion of nil into String 
    /usr/local/lib/ruby/3.0.0/json/common.rb:216 initialize
    /usr/local/lib/ruby/3.0.0/json/common.rb:216 new
    /usr/local/lib/ruby/3.0.0/json/common.rb:216 parse
    gems/sidekiq-6.2.2/lib/sidekiq.rb:180 load_json
    gems/sidekiq-pro-5.2.4/lib/sidekiq/batch/status.rb:74 callbacks
    gems/sidekiq-pro-5.2.4/lib/sidekiq/batch/callback.rb:117 complete
    gems/sidekiq-pro-5.2.4/lib/sidekiq/batch/callback.rb:15 perform
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:196 execute_job
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:164 block (2 levels) in process
    gems/sidekiq-6.2.2/lib/sidekiq/middleware/chain.rb:138 block in invoke
    gems/sidekiq-pro-5.2.4/lib/sidekiq/batch/middleware.rb:54 call
    gems/sidekiq-6.2.2/lib/sidekiq/middleware/chain.rb:140 block in invoke
    gems/sidekiq-ent-2.2.3/lib/sidekiq-ent/unique.rb:152 call
    gems/sidekiq-6.2.2/lib/sidekiq/middleware/chain.rb:140 block in invoke
    gems/sidekiq-ent-2.2.3/lib/sidekiq-ent/limiter/middleware.rb:38 call
    gems/sidekiq-6.2.2/lib/sidekiq/middleware/chain.rb:140 block in invoke
    gems/bugsnag-6.22.1/lib/bugsnag/integrations/sidekiq.rb:24 call
    gems/sidekiq-6.2.2/lib/sidekiq/middleware/chain.rb:140 block in invoke
    gems/sidekiq-6.2.2/lib/sidekiq/middleware/chain.rb:143 invoke
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:163 block in process
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:136 block (6 levels) in dispatch
    gems/sidekiq-6.2.2/lib/sidekiq/job_retry.rb:112 local
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:135 block (5 levels) in dispatch
    gems/sidekiq-6.2.2/lib/sidekiq/rails.rb:14 block in call
    gems/activesupport-6.1.7.2/lib/active_support/execution_wrapper.rb:91 wrap
    gems/activesupport-6.1.7.2/lib/active_support/reloader.rb:72 block in wrap
    gems/activesupport-6.1.7.2/lib/active_support/execution_wrapper.rb:91 wrap
    gems/activesupport-6.1.7.2/lib/active_support/reloader.rb:71 wrap
    gems/sidekiq-6.2.2/lib/sidekiq/rails.rb:13 call
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:131 block (4 levels) in dispatch
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:257 stats
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:126 block (3 levels) in dispatch
    gems/sidekiq-6.2.2/lib/sidekiq/job_logger.rb:13 call
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:125 block (2 levels) in dispatch
    gems/sidekiq-6.2.2/lib/sidekiq/job_retry.rb:79 global
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:124 block in dispatch
    gems/sidekiq-6.2.2/lib/sidekiq/logger.rb:11 with
    gems/sidekiq-6.2.2/lib/sidekiq/job_logger.rb:33 prepare
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:123 dispatch
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:162 process
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:78 process_one
    gems/sidekiq-6.2.2/lib/sidekiq/processor.rb:68 run
    gems/sidekiq-6.2.2/lib/sidekiq/util.rb:43 watchdog
    gems/sidekiq-6.2.2/lib/sidekiq/util.rb:52 block in safe_thread
    gems/logging-2.3.0/lib/logging/diagnostic_context.rb:474 block in create_with_logging_context
```

For more info, I'm running:

* sidekiq 6.2.2
* sidekiq-pro 5.2.4
* sidekiq-ent 2.2.3

I know we're not on the latest versions. We have to make some changes to allow us to upgrade.

**Comment by mperham** [comment]
If you click "1 week ago", it'll show you more specifics about that batch, any pending or dead PIDs, error backtraces, etc. Job loss can cause batches to back up. Are you seeing any job recovery from super_fetch in your logs?

**Reply by dhughes** [reply]
These may be from before `super_fetch` was enabled. I wasn't aware of it until recently. I'll watch for that. Thanks.
